### PR TITLE
Add support of special WrappingSearchAsyncActionPhase so the onPhaseStart() will always be followed by onPhaseEnd() within AbstractSearchAsyncAction

### PR DESCRIPTION
**Pull request by reta**
### Description
Add support of special WrappingSearchAsyncActionPhase so the onPhaseStart() will always be followed by onPhaseEnd() within AbstractSearchAsyncAction

### Related Issues
Reverts https://github.com/opensearch-project/OpenSearch/pull/12035 and fixes 

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
> ❌ Gradle check result for [0e43513](https://github.com/opensearch-project/OpenSearch/commit/0e435130c80825cc0c2213167e674c4f62a4bac8): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33685/)

https://github.com/opensearch-project/OpenSearch/issues/9464
https://github.com/opensearch-project/OpenSearch/issues/8771

**Comment by reta** [comment]
> ❌ Gradle check result for [9f38f6f](https://github.com/opensearch-project/OpenSearch/commit/9f38f6f1d49b4aa32b58f2d460d3ad2db2cd07d3): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33696/)

```
* What went wrong:
Execution failed for task ':distribution:bwc:bugfix:buildBwcLinuxTar'.
> Process '/var/jenkins/workspace/gradle-check/search/distribution/bwc/bugfix/build/bwc/checkout-2.11/gradlew [-Dbuild.snapshot=true, -Dscan.tag.NESTED, --parallel, distribution:archives:linux-tar:assemble, --build-cache]' finished with non-zero exit value 1
```

**Comment by reta** [comment]
> ❌ Gradle check result for [23defeb](https://github.com/opensearch-project/OpenSearch/commit/23defeb16d5a7c9bd7ef342c4b894c9803fee594): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33704/)

https://github.com/opensearch-project/OpenSearch/issues/9191

**Comment by reta** [comment]
> ❌ Gradle check result for [23defeb](https://github.com/opensearch-project/OpenSearch/commit/23defeb16d5a7c9bd7ef342c4b894c9803fee594): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33707/)

https://github.com/opensearch-project/OpenSearch/issues/10762
https://github.com/opensearch-project/OpenSearch/issues/9191

**Comment by reta** [comment]
> ❌ Gradle check result for [23defeb](https://github.com/opensearch-project/OpenSearch/commit/23defeb16d5a7c9bd7ef342c4b894c9803fee594): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33713/)

https://github.com/opensearch-project/OpenSearch/issues/12303

File: server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java [hunk 1]
```diff
@@ -637,6 +641,12 @@ public SearchPhase getCurrentPhase() {
 
     private void setCurrentPhase(SearchPhase phase) {
         currentPhase = phase;
+        // The WrappingSearchAsyncActionPhase (see please CanMatchPreFilterSearchPhase as one example) is a special case
+        // of search phase that wraps SearchAsyncActionPhase as SearchPhase. The AbstractSearchAsyncAction manages own
+        // onPhaseStart / onPhaseFailure / OnPhaseDone callbacks and the wrapping SearchPhase is being abandoned
+        // (fe, has no onPhaseEnd callbacks called ever). To fix that, we would not send any notifications for this
+        // phase.
+        currentPhaseHasLifecycle = !(phase instanceof WrappingSearchAsyncActionPhase);
```
Comment (dzane17) [inline server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java:649]: Can this be `(phase instanceof WrappingSearchAsyncActionPhase == false)`?

File: server/src/main/java/org/opensearch/action/search/WrappingSearchAsyncActionPhase.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+class WrappingSearchAsyncActionPhase extends SearchPhase {
+    private final AbstractSearchAsyncAction<? extends SearchPhaseResult> action;
+
+    protected WrappingSearchAsyncActionPhase(AbstractSearchAsyncAction<? extends SearchPhaseResult> action) {
+        super(action.getName());
+        this.action = action;
+    }
+
+    @Override
+    public void run() {
```
Comment (dzane17) [inline server/src/main/java/org/opensearch/action/search/WrappingSearchAsyncActionPhase.java:28]: Should we move this `run()` method here? Right now `WrappingSearchAsyncActionPhase` is only used by `CanMatchPreFilterSearchPhase` but not sure if we need this to be the default behavior for ALL "wrapping" search phases.

Comment (reta) [reply server/src/main/java/org/opensearch/action/search/WrappingSearchAsyncActionPhase.java:28]: It should be behaviour for all: this class is a wrapper around  `AbstractSearchAsyncActionPhase` and has intrinsic knowledge how the wrapped class works. It "knows" exactly what start does.

File: server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java (hunk 1 above)
Comment (reta) [reply server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java:649]: Sure!

**Review by dzane17 — APPROVED** [review]

**Review by andrross — APPROVED** [review]

**Comment by reta** [comment]
> ❌ Gradle check result for [a231a63](https://github.com/opensearch-project/OpenSearch/commit/a231a6311543293cf4684108ef4b8ff0924a420e): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33724/)

https://github.com/opensearch-project/OpenSearch/issues/12308

**Comment by reta** [comment]
> ❌ Gradle check result for [a231a63](https://github.com/opensearch-project/OpenSearch/commit/a231a6311543293cf4684108ef4b8ff0924a420e): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/33730/)

https://github.com/opensearch-project/OpenSearch/issues/9191

**Comment by reta** [comment]
@andrross mind please re-approving (after conflict resolution), thank you!

**Comment by reta** [comment]
@andrross mind please re-approving (after conflict resolution), thank you!

**Pull request merged**